unittests/finality_proof.hpp

```diff
@@ -0,0 +1,289 @@
+
+      if (if_ext.new_finalizer_policy_diff.has_value()) return true;
+      else return false;
+
+   }
+
+   static eosio::chain::finalizer_policy update_finalizer_policy(const signed_block_ptr block, eosio::chain::finalizer_policy current_policy){
+
+      // extract new finalizer policy
+      instant_finality_extension if_ext = block->extract_header_extension<instant_finality_extension>();
+
+      assert(if_ext.new_finalizer_policy_diff.has_value());
+
+      eosio::chain::finalizer_policy active_finalizer_policy =
+         current_policy.apply_diff(if_ext.new_finalizer_policy_diff.value());
+
+      return active_finalizer_policy;
+
+   }
+
+   struct policy_count {
+      finalizer_policy policy;
+      int32_t blocks_since_proposed;
+   };
+
+   template<size_t NUM_NODES>
+   class proof_test_cluster : public finality_test_cluster<NUM_NODES> {
+   public:
+
+      /*****
+
+         The proof_test_cluster class inherits from finality_test_cluster and serves to generate finality proofs for the purpose of IBC and proving finality violations.
+
+         It has its own high-level produce_block function, which hides all the internal consensus details, and returns an extended struct containing data relevant for proof generation.
+
+         It doesn't support forks or rollbacks, and always assumes the happy path in finality progression, which is sufficient for the purpose of generating finality proofs for testing.
+         
+         It also assumes a single producer pre-transition, resulting in only 2 transition blocks when IF is activated.
+
+      *****/
+         
+      // cache last proposed, last pending and currently active finalizer policies + digests
+      eosio::chain::finalizer_policy last_proposed_finalizer_policy;
+      digest_type last_proposed_finalizer_policy_digest;
+
+      eosio::chain::finalizer_policy last_pending_finalizer_policy;
+      digest_type last_pending_finalizer_policy_digest;
+
+      eosio::chain::finalizer_policy active_finalizer_policy;
+      digest_type active_finalizer_policy_digest;
+
+      // counter to (optimistically) track internal policy changes
+      std::unordered_map<digest_type, policy_count> blocks_since_proposed_policy;
+
+      //node0 always produce blocks. This vector determines if votes from node1, node2, etc. are propagated
+      std::vector<bool> vote_propagation;
+      
+      // internal flag to indicate whether or not block is the IF genesis block
+      bool is_genesis = true;
+      // internal flag to indicate whether or not the transition is complete
+      bool is_transition = true;
+
+      // returns finality leaves for construction of merkle proofs
+      std::vector<digest_type> get_finality_leaves(const size_t cutoff){
+         assert(cutoff>=0 && cutoff<finality_leaves.size());
+         return std::vector<digest_type>(finality_leaves.begin(), finality_leaves.begin() + cutoff + 1);

```

Comment:
Personally I prefer not repeating types if not needed:
```suggestion
         return { finality_leaves.begin(), finality_leaves.begin() + cutoff + 1 };

```